src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTrackerGeneratorInfo.cs

```diff
@@ -35,62 +33,27 @@ private partial class CompilationTracker
 
-            /// <summary>
-            /// Whether the generated documents are frozen and generators should never be ran again, ever, even if a document
-            /// is later changed. This is used to ensure that when we produce a frozen solution for partial semantics,
-            /// further downstream forking of that solution won't rerun generators. This is because of two reasons:
-            /// <list type="number">
-            /// <item>Generally once we've produced a frozen solution with partial semantics, we now want speed rather
-            /// than accuracy; a generator running in a later path will still cause issues there.</item>
-            /// <item>The frozen solution with partial semantics makes no guarantee that other syntax trees exist or
-            /// whether we even have references -- it's pretty likely that running a generator might produce worse results
-            /// than what we originally had.</item>
-            /// </list>
-            /// </summary>
-            public readonly bool DocumentsAreFinalAndFrozen;
-
             public CompilationTrackerGeneratorInfo(
                 TextDocumentStates<SourceGeneratedDocumentState> documents,
-                GeneratorDriver? driver,
-                bool documentsAreFinal) : this(documents, driver, documentsAreFinal, documentsAreFinalAndFrozen: false)
+                GeneratorDriver? driver) : this(documents, driver, documentsAreFinal: false)
             {
             }
 
             private CompilationTrackerGeneratorInfo(
                 TextDocumentStates<SourceGeneratedDocumentState> documents,
                 GeneratorDriver? driver,
-                bool documentsAreFinal,
-                bool documentsAreFinalAndFrozen)
+                bool documentsAreFinal)
             {
                 Documents = documents;
                 Driver = driver;
                 DocumentsAreFinal = documentsAreFinal;
-                DocumentsAreFinalAndFrozen = documentsAreFinalAndFrozen;
-
-                // If we're frozen, that implies final as well
-                Contract.ThrowIfTrue(documentsAreFinalAndFrozen && !documentsAreFinal);
             }
 
             public CompilationTrackerGeneratorInfo WithDocumentsAreFinal(bool documentsAreFinal)
-            {
-                // If we're already frozen, then we won't do anything even if somebody calls WithDocumentsAreFinal(false);
-                // this for example would happen if we had a frozen snapshot, and then we fork it further with additional changes.
-                // In that case we would be calling WithDocumentsAreFinal(false) to force generators to run again, but if we've
-                // frozen in partial semantics, we're done running them period. So we'll just keep treating them as final,
-                // no matter the wishes of the caller.
-                if (DocumentsAreFinalAndFrozen || DocumentsAreFinal == documentsAreFinal)
```

Comment:
i always found thsi logic here confusing to have to reason about.  now it is extracted to the one place that woudl actually want to set doucmentsAreFinal=false.